src/service/xhr-impl.js

```diff
@@ -186,31 +184,28 @@ export class Xhr {
     if (!this.ampdocSingle_) {
       return Promise.resolve();
     }
-
-    const htmlElement = this.ampdocSingle_.getRootNode().documentElement;
-    const docOptedIn = htmlElement.hasAttribute('allow-xhr-interception');
-    if (!docOptedIn) {
-      return Promise.resolve();
-    }
-
     const viewer = Services.viewerForDoc(this.ampdocSingle_);
-    if (!viewer.hasCapability('xhrInterceptor')) {
-      return Promise.resolve();
-    }
-
-    return viewer.isTrustedViewer().then(viewerTrusted => {
-      if (!viewerTrusted && !getMode(this.win).development) {
-        return;
-      }
-
-      const messagePayload = dict({
-        'originalRequest': this.toStructuredCloneable_(input, init),
-      });
-
-      return viewer.sendMessageAwaitResponse('xhr', messagePayload)
-          .then(response =>
-            this.fromStructuredCloneable_(response, init.responseType));
-    });
+    return viewer.whenFirstVisible()
+        .then(() => {
+          const htmlElement = this.ampdocSingle_.getRootNode().documentElement;
+          const docOptedIn = htmlElement.hasAttribute('allow-xhr-interception');
+          if (!docOptedIn || !viewer.hasCapability('xhrInterceptor')) {
+            return Promise.resolve();
+          }
+          return viewer.isTrustedViewer();
+        }).then(viewerTrusted => {
+          if (!viewerTrusted && !getMode(this.win).development) {
```

Comment:
Sorry, this lead to confusion when I recommended unnesting. What I meant is that these should be two separate promise chains:

```js
// Shorthand, don't actually write like this
maybeIntercept_(input, init) {
  if (!single) return Promise.resolve();

  const vis = viewer.whenFirstVisible();
  if (!hasCapability) return vis;

  return vis.then(() => {
    if (!optedIn) {
      return false;
    }
    return viewer.isTrusted()
  }).then(trusted => {
    if (!trusted) return;
    return sendMessage(...);
  });
}

Reply:
@jridgewell  no worries. hopefully i got it right this time. otherwise we can discuss in person. thanks